## Setup

In [1]:
from main import RandomPlayer
from quarto.players import MyPlayer
from quarto.objects import Quarto
from IPython.display import clear_output

In [2]:
def main(reverse = False):
    game = Quarto()
    if reverse:
        game.set_players((RandomPlayer(game), MyPlayer(game))) 
    else:
        game.set_players((MyPlayer(game), RandomPlayer(game)))
    
    winner = game.run()
    clear_output(wait=False)
    print(f"main: Winner: player {winner}")
    return winner

### Tournament

In [3]:
N_MATCHES = 200

cnt = 0
for i in range(N_MATCHES):
    if i < N_MATCHES//2:  
        w = main()
        if w == 0:
            cnt += 1
    else:
        w = main(True)
        if w == 1:
            cnt += 1
    

print(f'WINRATE = {cnt / N_MATCHES}') 

main: Winner: player 1
WINRATE = 0.965


| # matches | delay_minmax = True | WinRate | Execution time | ExecutionTime / # Matches    |
|  :-----:  |   :---------:       |  :---:  | :---------:    | :------:    |
| 20        |       True          |   0.95  |   6m 45.8s     |  20.25s                   |
| 20        |       True          |   0.9  |   5m 14.2s     | 15.7s                     |
| 20        |      True           |   1.00  |  1m 39.8s      | 4.95s                     |
| 20        |      False          |   0.95  |    5m 52.0s    | 17.6s                     |
| 20        |      False          |   1.00  |    6m 59.4s    | 20.95s                    |
| 20        |      False          |   0.9   |   11m 10.8s    | 33.5s                     |
| 100       |      True           |   0.92  | 6m 21.2s       | 3.81s                     |
| 100       |      True           |   0.91  |  3m 24.5s      | 2.04s                     |
| 100       |     False           |  0.91   |   3m 24.5s     | 2.04s                     |  
| 100       |     False           |  0.96   |  12m 14.0s     | 7.34s                     |
| 200       |     False           |  0.975  |   37m 37.5s    | 11.29s                    |
| 200       |     Ture            |  0.965  |     13m 1.4s   | 3.91s                     |


| delay_minmax | Avg WinRate | Avg ExecutionTime/#matches |
|  :-----:     |   :-----:   |  :-----------:             |
|   True       |   0.924     |    8.44s                   |
|   False      |   0.949     |    15.45s                  |